sys/net/gnrc/network_layer/ipv6/nib/_nib-slaac.c

```diff
@@ -0,0 +1,204 @@
+                          uint8_t pfx_len)
+{
+    ipv6_addr_t addr = IPV6_ADDR_UNSPECIFIED;
+    int idx;
+    uint8_t flags = GNRC_NETIF_IPV6_ADDRS_FLAGS_STATE_TENTATIVE;
+
+    DEBUG("nib: add address based on %s/%u automatically to interface %u\n",
+          ipv6_addr_to_str(addr_str, pfx, sizeof(addr_str)),
+          pfx_len, netif->pid);
+#if GNRC_IPV6_NIB_CONF_6LN
+    bool new_address = false;
+#endif  /* GNRC_IPV6_NIB_CONF_6LN */
+    gnrc_netif_ipv6_get_iid(netif, (eui64_t *)&addr.u64[1]);
+    ipv6_addr_init_prefix(&addr, pfx, pfx_len);
+    if ((idx = gnrc_netif_ipv6_addr_idx(netif, &addr)) < 0) {
+        if ((idx = gnrc_netif_ipv6_addr_add_internal(netif, &addr, pfx_len,
+                                                     flags)) < 0) {
+            DEBUG("nib: Can't add link-local address on interface %u\n",
+                  netif->pid);
+            return;
+        }
+#if GNRC_IPV6_NIB_CONF_6LN
+        new_address = true;
+#endif  /* GNRC_IPV6_NIB_CONF_6LN */
+    }
+
+#if GNRC_IPV6_NIB_CONF_6LN
+    /* mark link-local addresses as valid on 6LN */
+    if (gnrc_netif_is_6ln(netif) && ipv6_addr_is_link_local(pfx)) {
+        /* don't do this beforehand or risk a deadlock:
+         *  - gnrc_netif_ipv6_addr_add_internal() adds VALID (i.e. manually configured
+         *    addresses to the prefix list locking the NIB's mutex which is already
+         *    locked here) */
+        netif->ipv6.addrs_flags[idx] &= ~GNRC_NETIF_IPV6_ADDRS_FLAGS_STATE_MASK;
+        netif->ipv6.addrs_flags[idx] |= GNRC_NETIF_IPV6_ADDRS_FLAGS_STATE_VALID;
+    }
+#endif  /* GNRC_IPV6_NIB_CONF_6LN */
+#if GNRC_IPV6_NIB_CONF_6LN
+    if (new_address && gnrc_netif_is_6ln(netif) &&
+        !gnrc_netif_is_6lbr(netif)) {
+        _handle_rereg_address(&netif->ipv6.addrs[idx]);
+    }
+#else   /* GNRC_IPV6_NIB_CONF_6LN */
+    (void)idx;
+#endif  /* GNRC_IPV6_NIB_CONF_6LN */
+}
+#endif  /* GNRC_IPV6_NIB_CONF_6LN || GNRC_IPV6_NIB_CONF_SLAAC */
+
+#if GNRC_IPV6_NIB_CONF_SLAAC
+static bool _try_l2addr_reconfiguration(gnrc_netif_t *netif)
+{
+    uint8_t hwaddr[GNRC_NETIF_L2ADDR_MAXLEN];
+    uint16_t hwaddr_len;
+
+    if (gnrc_netapi_get(netif->pid, NETOPT_SRC_LEN, 0, &hwaddr_len,
+                        sizeof(hwaddr_len)) < 0) {
+        return false;
+    }
+    luid_get(hwaddr, hwaddr_len);
+#if GNRC_IPV6_NIB_CONF_6LN
+    if (hwaddr_len == IEEE802154_LONG_ADDRESS_LEN) {
+        if (gnrc_netapi_set(netif->pid, NETOPT_ADDRESS_LONG, 0, hwaddr,
+                            hwaddr_len) < 0) {
+            return false;
+        }
+    }
+    else
+#endif
+    if (gnrc_netapi_set(netif->pid, NETOPT_ADDRESS, 0, hwaddr,
+                        hwaddr_len) < 0) {
+        return false;
+    }
+    return true;
+}
+
+static bool _try_addr_reconfiguration(gnrc_netif_t *netif)
+{
+    eui64_t orig_iid;
+    bool remove_old = false, hwaddr_reconf;
+
+    if (gnrc_netif_ipv6_get_iid(netif, &orig_iid) == 0) {
```

Comment:
this function should only be run, if hwaddr_reconf is true, so I'd suggest to move it into the if statement below. Also can't we get rid of both bools by moving the respective functions into the if conditions and make the code more readable that way?

Reply:
> this function should only be run, if hwaddr_reconf is true, so I'd suggest to move it into the if statement below. 

If `hwaddr_reconf` is true, the link-layer address was changed (that what the name of this variable implies :wink:), so the IID would also been changed => we can't fetch the original IID (`orig_iid`) anymore => It *has* to be called before `_try_l2addr_reconfiguration()` is called (which determines if `hwaddr_reconf` is true).

> Also can't we get rid of both bools by moving the respective functions into the if conditions and make the code more readable that way?

`hwaddr_reconf` is required, since we need to release the interface again before the if-statement. Likewise, we need `remove_old`, because as mentioned above, the call order of the functions is important.

Reply:
> we can't fetch the original IID (`orig_iid`) anymore

We need that to determine, if the IP address is based on the old link-layer address, to remove it, in case that wasn't clear.

Reply:
>the call order of the functions is important.

ah, okay wasn't aware of that ...